rust/osm2lanes/src/transform/tags_to_lanes/road.rs

```diff
@@ -0,0 +1,412 @@
+    fn strength(&self) -> Hunch<()> {
+        match self.hunch() {
+            Some(h) => h.strength(),
+            None => Guessed(()),
+        }
+    }
+}
+
+// trait HasProblems { fn iter<Problem>... }
+// type TheoryHistory = Vec<Problem>;
+struct Theory<'a, T> {
+    history: Vec<(Hunch<T>, Vec<&'a Theory<'a, T>>)>,
+    // history: Vec<(Hunch<T>, &impl(HasProblems + HunchStrength))>
+    problems: Option<Vec<Problem>>,
+}
+impl<'a, T> Theory<'a, T> {
+    fn hunch(self) -> Option<&'a Hunch<T>> {
+        match self.history {
+            [] => None,
+            [.., (h, _)] => Some(h),
+        }
+    }
+
+    fn update(
+        &mut self,
+        new_hunch: Hunch<T>,
+        supporting_theories: Vec<&Theory<T>>,
+        transform_err: Option<fn() -> Problem>,
+    ) /*-> Result<(), ()>*/
+    {
+        self.history.push((
+            if let Some(existing_hunch) = self.hunch() {
+                new_hunch.reconcile(existing_hunch, || {
+                    self.problems.push(match transform_err {
+                        Some(f) => f(),
+                        None => "conflict",
+                    })
+                })
+            } else {
+                new_hunch
+            },
+            supporting_theories,
+        ))
+    }
+
+    fn new(new_hunch: Hunch<T>, supporting_theories: Vec<&Theory<T>>) -> Self {
+        let mut s = Self::default();
+        s.update(new_hunch, supporting_theories, None);
+        s
+    }
+
+    fn default() -> Self {
+        Self {
+            history: Vec::default(),
+            problems: None,
+        }
+    }
+    fn from(hunch: Hunch<T>) -> Self {
+        Self {
+            history: vec![(hunch, Vec::default())],
+            problems: None,
+        }
+    }
+}
+
+impl<'a, T: FromStr> Theory<'a, T> {
+    // Convenience
+    fn update_with_tag(&mut self, tag_str: Option<&str>) {
+        if let Some(tag) = tag_str {
+            match tag.parse() {
+                Ok(val) => self.update(Tagged(val), Vec::default(), None),
+                Err(e) => self.problems.push(e.description()),
+            }
+        }
+    }
+}
+
+/// A set of Hunches that describe canonically the state of a schema.
+trait Scheme {
+    /// Get a scheme with all the base Assumptions.
+    fn default_for_road(road: &Road) -> Self;
+    /// Set tagged values and update conflicts.
+    fn set_tags(&self, tags: &Tags);
+    /// Spruce it up with Guesses, make it look plausible, given what we know.
+    fn guess(&self, road: &Road);
+
+    // fn problems(&self) -> Option<Vec<Problem>>;
+}
+
+enum Oneway {
+    Yes,
+    Reverse,
+    No,
+}
+
+struct Lanes<'a> {
+    lanes: Theory<'a, Count>,
+    forward: Theory<'a, Count>,
+    backward: Theory<'a, Count>,
+    bothways: Theory<'a, Count>,
+    // directions: Theory<'a, Vec<LaneDirection>>,
+}
+
+struct Busway<'a> {
+    forward: Theory<'a, Option<LaneDirection>>,
+    backward: Theory<'a, Option<LaneDirection>>,
+}
```

Comment:
having so many lifetime specifiers usually indicates a code smell in rust FYI, as I don't see why you should need them.

Reply:
Yep. I think something forces me to include it, and rust needs assurance here, but I haven't looked into it yet. Hopefully its not needed in the end!